File: Xamarin/SampleXamarin.iOS/Views/SensorStatusView.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using CoreGraphics;
using Foundation;
using ObjCRuntime;
using System;
using UIKit;

namespace SampleXamarin.iOS
{
    [Register(nameof(SensorStatusView))]
    public class SensorStatusView : UIView
    {
        [Outlet] private UIImageView GeoLocationStatusIcon { get; set; }
        [Outlet] private UIImageView WifiStatusIcon { get; set; }
        [Outlet] private UIImageView BluetoothStatusIcon { get; set; }

        public SensorStatusModel Model { private get; set; }

        public SensorStatusView(IntPtr handle)
            : base(handle)
        {
            Initialize();
        }

        public SensorStatusView(CGRect frame)
            : base(frame)
        {
            Initialize();
        }

        public SensorStatusView(NSCoder aDecoder)
            : base(aDecoder)
        {
            Initialize();
        }

        private void Initialize()
        {
            var ownType = typeof(SensorStatusView);
            var bundle = NSBundle.FromClass(new Class(ownType));
            var nib = bundle.LoadNib(ownType.Name, owner: this, options: null);
            var contentView = nib?.GetItem<UIView>(0);
            if (contentView != null)
            {
                var adjustedFrame = Frame;
                adjustedFrame.Size = contentView.Frame.Size;
                Frame = adjustedFrame;
                AddSubview(contentView);
            }
        }

        public void Update()
        {
            GeoLocationStatusIcon.Image = GetStatusIcon(status: Model?.GeoLocationStatus);
            WifiStatusIcon.Image = GetStatusIcon(status: Model?.WifiSignalStatus);
            BluetoothStatusIcon.Image = GetStatusIcon(status: Model?.BluetoothSignalStatus);
        }

        private UIImage GetStatusIcon(SensorStatus? status)
        {
            switch (status ?? SensorStatus.Indeterminate)
            {
                case SensorStatus.Indeterminate: return UIImage.FromBundle("gray-circle");
                case SensorStatus.Blocked: return UIImage.FromBundle("red-circle");
                case SensorStatus.Unavailable: return UIImage.FromBundle("orange-circle");
                case SensorStatus.Available: return UIImage.FromBundle("green-circle");
            }

            throw new InvalidOperationException("Invalid sensor status");
        }
    }
}

File: iOS/Swift/SampleSwift/Views/SensorStatusView.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

import Foundation

class SensorStatusView: UIView {
    @IBOutlet var geoLocationStatusIcon: UIImageView!
    @IBOutlet var wifiStatusIcon: UIImageView!
    @IBOutlet var bluetoothStatusIcon: UIImageView!

    private var model: SensorStatusModel?

    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }

    func setModel(_ model: SensorStatusModel?) {
        self.model = model
    }

    private func commonInit() {
        let ownType = type(of: self)
        let bundle = Bundle(for: ownType)
        let nib = bundle.loadNibNamed(String(describing: ownType), owner: self, options: nil)
        if let contentView = nib?.first as? UIView {
            self.frame.size = contentView.frame.size
            addSubview(contentView)
        }
    }

    public func update() {
        geoLocationStatusIcon.image = getStatusIcon(status: model?.geoLocationStatus)
        wifiStatusIcon.image = getStatusIcon(status: model?.wifiSignalStatus)
        bluetoothStatusIcon.image = getStatusIcon(status: model?.bluetoothSignalStatus)
    }

    private func getStatusIcon(status: SensorStatus?) -> UIImage? {
        switch status ?? .Indeterminate {
        case .Indeterminate: return UIImage(named: "gray-circle")
            case .Blocked: return UIImage(named: "red-circle")
            case .Unavailable: return UIImage(named: "orange-circle")
            case .Available: return UIImage(named: "green-circle")
        }
    }
}
